projects/serializers/section.py

```diff
@@ -1,64 +1,38 @@
-from collections import namedtuple
+import collections
 
-from django.contrib.auth import get_user_model
-from django.forms import DateInput
 from rest_framework import serializers
 
-from projects.models import Attribute
+from projects.serializers.utils import (
+    _get_serializer_field_data,
+    _is_attribute_required,
+)
 
 
-class DateInputType(DateInput):
-    input_type = "date"
+def create_section_serializer(section, context):
+    """
+    Dynamically create a serializer for a ProjectPhaseSection instance
 
+    Since a attribute of a section can be dynamically defined, there is
+    no way of knowing which fields should be in the serializer before
+    checking the current state of the section.
 
-FIELD_TYPES = {
-    Attribute.TYPE_SHORT_STRING: serializers.CharField,
-    Attribute.TYPE_LONG_STRING: serializers.CharField,
-    Attribute.TYPE_INTEGER: serializers.IntegerField,
-    Attribute.TYPE_BOOLEAN: serializers.BooleanField,
-    Attribute.TYPE_DATE: serializers.DateField,
-    Attribute.TYPE_IMAGE: serializers.ImageField,  # TODO: Figure out file uploads with DRF
-    Attribute.TYPE_USER: serializers.PrimaryKeyRelatedField,
-    # TODO Add Attribute.TYPE_GEOMETRY
-}
+    Additionally, the incoming request is also checked in order to
+    get the relevant fields for that exact request. This has to be done
+    as the serialization not only relies on the section instance but
+    also on the input data field values and their relationship with
+    other fields values.
+    """
 
-FieldData = namedtuple("FieldData", ["field_class", "field_arguments"])
+    request = context.get("request", None)
+    attribute_data = get_attribute_data(request)
 
+    if not request:
+        return None
 
-def _get_serializer_field_data(attribute):
-    field_arguments = {}
-    field_class = FIELD_TYPES.get(attribute.value_type, None)
-
-    choices = attribute.value_choices.all()
-    if choices:
-        field_arguments["choices"] = choices
-
-        if attribute.multiple_choice:
-            field_class = serializers.MultipleChoiceField
-
-    if attribute.value_type == Attribute.TYPE_USER:
-        field_arguments["queryset"] = get_user_model().objects.all()
-
-    field_arguments["help_text"] = attribute.help_text
-
-    return FieldData(field_class, field_arguments)
-
-
-def _is_attribute_required(section_attribute):
-    attribute = section_attribute.attribute
-
-    if (
-        attribute.value_type != Attribute.TYPE_BOOLEAN
-        and not section_attribute.generated
-    ):
-        return section_attribute.required
-    else:
-        return False
-
-
-def create_section_serializer(section):
     serializer_fields = {}
     for section_attribute in section.projectphasesectionattribute_set.order_by("index"):
+        if not is_relevant_attribute(section_attribute, attribute_data):
```

Comment:
Just thinking if this would handle a situation where the value for the field referenced by `relies_on` is already set and that field might not be sent in the request. Will the value for the current `section_attribute` be considered as relevant in that case?

Reply:
While this line itself did not change and therefore is not marked as "Outdated", I've now updated the PR so that existing attribute data is included in the relevancy check :) Good catch indeed!